connectors/servicenow/src/main/java/com/vmware/connectors/servicenow/ServiceNowController.java

```diff
@@ -159,7 +166,7 @@ public ServiceNowController(
                 .header(AUTHORIZATION, auth)
                 .retrieve()
                 .bodyToMono(JsonDocument.class)
-                .map(userInfoResponse -> userInfoResponse.read("$.result[0]." + SysUser.Fields.SYS_ID));
+                .map(userInfoResponse -> Optional.ofNullable(userInfoResponse.read("$.result[0]." + SysUser.Fields.SYS_ID)));
```

Comment:
I didn't see a way to have a null turn into an empty.  Returning null from map throws a NPE.  None of the methods in Mono seemed to do what I wanted: https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html

Reply:
Unless maybe I do a flatMap and return a mono of the result (one of the ones that takes in a null -> empty).  That seemed weird, but I can try it if you think it's better.

Reply:
```
                .bodyToMono(JsonDocument.class)
                .flatMap(userInfoResponse -> {
                    String userSysId = userInfoResponse.read("$.result[0]." + SysUser.Fields.SYS_ID);
                    if (userSysId == null) {
                        logger.warn("sys_id for {} not found in {}, returning empty cards", email, baseUrl);
                    }
                    return Mono.justOrEmpty(userSysId);
                });
```

This loses us the Request ID in the logs:

```
2018-04-06 17:52:15,894 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x2dbb5091, L:/10.0.2.15:45210 - R:dev15329.service-now.com/103.23.67.118:443] Writing object DefaultHttpRequest(decodeResult: success, version: HTTP/1.1)
GET /api/now/table/sys_user?sysparm_fields=sys_id&sysparm_limit=1&email=mem1@vmware.com HTTP/1.1
user-agent: ReactorNetty/0.7.5.RELEASE
host: dev15329.service-now.com
accept: */*
accept-encoding: gzip
Authorization: Bearer <token>
2018-04-06 17:52:16,370 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x2dbb5091, L:/10.0.2.15:45210 - R:dev15329.service-now.com/103.23.67.118:443] Writing object EmptyLastHttpContent
2018-04-06 17:52:16,813 WARN [reactor-http-nio-4]   c.v.c.s.ServiceNowController [ServiceNowController.java:165] sys_id for mem1@vmware.com not found in https://dev15329.service-now.com, returning empty cards
2018-04-06 17:52:16,817 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x2dbb5091, L:/10.0.2.15:45210 - R:dev15329.service-now.com/103.23.67.118:443] Disposing context reactor.ipc.netty.channel.PooledClientContextHandler@67beb405
```

Is there a way for me to keep the context?

Reply:
```
                .bodyToMono(JsonDocument.class)
                .flatMap(Reactive.wrapFlatMapper(userInfoResponse -> {
                    String userSysId = userInfoResponse.read("$.result[0]." + SysUser.Fields.SYS_ID);
                    if (userSysId == null) {
                        logger.warn("sys_id for {} not found in {}, returning empty cards", email, baseUrl);
                    }
                    return Mono.justOrEmpty(userSysId);
                }));
```

This works because callForUserSysId is called within the same thread as the getCards method:

```
2018-04-06 18:04:25,327 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x5baa9fd5, L:/10.0.2.15:45218 - R:dev15329.service-now.com/103.23.67.118:443] Writing object DefaultHttpRequest(decodeResult: success, version: HTTP/1.1)
GET /api/now/table/sys_user?sysparm_fields=sys_id&sysparm_limit=1&email=mem1@vmware.com HTTP/1.1
user-agent: ReactorNetty/0.7.5.RELEASE
host: dev15329.service-now.com
accept: */*
accept-encoding: gzip
Authorization: Bearer <token>
2018-04-06 18:04:25,903 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x5baa9fd5, L:/10.0.2.15:45218 - R:dev15329.service-now.com/103.23.67.118:443] Writing object EmptyLastHttpContent
2018-04-06 18:04:26,469 WARN [reactor-http-nio-4] B46BE703-307E-4703-AA27-035DDFC91643 bardj@HEROCARD c.v.c.s.ServiceNowController [ServiceNowController.java:165] sys_id for mem1@vmware.com not found in https://dev15329.service-now.com, returning empty cards
2018-04-06 18:04:26,473 DEBUG [reactor-http-nio-4]   r.i.n.c.ChannelOperationsHandler [Loggers.java:218] [id: 0x5baa9fd5, L:/10.0.2.15:45218 - R:dev15329.service-now.com/103.23.67.118:443] Disposing context reactor.ipc.netty.channel.PooledClientContextHandler@1a4f7611
```

This seems pretty error prone to me.  It seems like it's a refactor away from being in the wrong thread and not working anymore.